File: impl/python/src/img_drv/edsl.py
"""The eDSL surface: DESCRIBE a build, and get a derivation back.

Everything else in this library reads derivations. This module is the other
direction, and it is the half of the thesis that can actually fail: if the
first-order signature in ``docs/spec/signature.md`` is enough, then a build is
described by a single application of :func:`derivation`, and the host language
supplies everything else.

    >>> from img_drv import derivation
    >>> hello = derivation(
    ...     name="hello",
    ...     system="x86_64-linux",
    ...     builder="/bin/sh",
    ...     args=["-c", "echo hi > $out"],
    ... )
    >>> hello.output()
    '/nix/store/mjs27ix6ig2bkbi3s3sm470vrv4lf7ic-hello'

That path is Nix's, byte for byte, and it is known BEFORE anything is built.
Depending on a derivation means naming the outputs you need:

    >>> dependent = derivation(
    ...     name="dependent",
    ...     system="x86_64-linux",
    ...     builder="/bin/sh",
    ...     args=["-c", f"cat {hello.output()} > $out"],
    ...     input_drvs=[hello],
    ... )

Note that the reference in ``args`` and the edge in ``input_drvs`` are written
separately. Nix couples them through string contexts, which is a feature of its
EVALUATOR; the signature deliberately has no computation in it, so the coupling
is the caller's to make. That is the cost of the design, and stating it is more
honest than hiding it behind an interpolation helper.

What this module adds over the raw :class:`~img_drv.derivation.Derivation`
record is exactly two things: the CANONICAL form (which orderings are
load-bearing), and the KNOT (output paths are the hash of the derivation that
contains them). Both are the sort of thing that produces a plausible wrong
answer rather than an error, which is why both are checked against real Nix
output rather than against examples.
"""

from __future__ import annotations

import base64
import binascii
import json
import pathlib
import string
from collections.abc import Mapping, Sequence
from dataclasses import dataclass, replace
from typing import Final, Literal, TypeAlias, cast

from .aterm import unparse
from .derivation import (
    Derivation,
    InputDrv,
    Output,
    OutputName,
    Sha256Hex,
    StorePath,
)
from .store import (
    base32_decode,
    base32_length,
    drv_path,
    fixed_output_input_hash,
    output_paths,
    sha256_hex,
)

__all__ = [
    "Dep",
    "Drv",
    "FixedOutput",
    "HashAlgo",
    "HashMode",
    "InvalidDerivationError",
    "JsonValue",
    "canonical",
    "derivation",
    "json_env",
    "valid_name",
]

#: A JSON value, as `__structuredAttrs` carries it.
#:
#: This is the first RECURSIVE type in the signature. Everything else is a
#: product of primitives and lists of them, which is what `docs/theory.md`
#: section 1 restricts the signature to; a JSON value is an inductive datatype,
#: a fixed point of a polynomial functor. Still first-order and still
#: algebraic, so the Lawvere argument survives, but it is a genuine extension
#: and it is where a language without sum types has to work hardest.
#: Written as a `TypeAlias` with forward references rather than the 3.12
#: `type` statement, because the supported floor is 3.11.
JsonValue: TypeAlias = (
    str
    | int
    | float
    | bool
    | Sequence["JsonValue"]
    | Mapping[str, "JsonValue"]
    | None
)

#: The hash algorithms Nix accepts for a fixed-output derivation.
HashAlgo = Literal["md5", "sha1", "sha256", "sha512"]

#: How the output is ingested: a single file, or a NAR of a directory tree.
#: ``recursive`` is what puts the ``r:`` prefix on the serialized algorithm,
#: and ``r:sha256`` selects an entirely different store-path scheme.
HashMode = Literal["flat", "recursive"]

_DIGEST_BYTES: Final[Mapping[str, int]] = {
    "md5": 16,
    "sha1": 20,
    "sha256": 32,
    "sha512": 64,
}

#: Nix's store-name character set, and its length cap.
_NAME_CHARS: Final[frozenset[str]] = frozenset(
    string.ascii_letters + string.digits + "+-._?="
)
_NAME_MAX: Final = 211

#: Env keys this module derives from the other arguments. Passing one directly
#: would let the env disagree with the field it mirrors, which is a derivation
#: Nix would never emit.
_HASH_KEYS: Final[frozenset[str]] = frozenset(
    ("outputHash", "outputHashAlgo", "outputHashMode")
)
_ALWAYS_RESERVED: Final[frozenset[str]] = (
    frozenset(
        ("name", "system", "builder", "outputs", "__json", "__structuredAttrs")
    )
    | _HASH_KEYS
)

_DEFAULT_OUTPUT: Final = OutputName("out")


class InvalidDerivationError(ValueError):
    """A description that violates an invariant in ``spec/signature.md``.

    Raised at CONSTRUCTION time. The alternative is a derivation that
    serializes perfectly and means something else, which this project has
    already paid for once.
    """


def valid_name(name: str) -> bool:
    """Whether ``name`` is usable as a store path name.

    Accepts every name in the real corpus. That shows the predicate is not too
    strict; it does not show it is not too permissive, which is why
    ``spec/signature.md`` still lists the exact rules as open.
    """
    return (
        0 < len(name) <= _NAME_MAX
        and name not in (".", "..")
        and not name.startswith(".")
        and not set(name) - _NAME_CHARS
    )


def _as_algo(text: str) -> HashAlgo:
    if text not in _DIGEST_BYTES:
        raise InvalidDerivationError(
            f"unknown hash algorithm {text!r}; "
            f"expected one of {sorted(_DIGEST_BYTES)}"
        )
    return cast(HashAlgo, text)


@dataclass(frozen=True, slots=True)
class FixedOutput:
    """A declared result: the derivation's identity comes from this hash.

    ``hash`` is kept EXACTLY as written, because that is what reaches the env,
    while the outputs tuple carries it re-encoded as hex. Both forms are in
    ``examples/fixed.drv``, and the rule is verified on all 93 fixed-output
    derivations in the real corpus.

    ``algo`` may be omitted when ``hash`` is SRI (``sha256-<base64>``), which
    already names its algorithm. Real derivations do exactly that: 11 of the 93
    carry no ``outputHashAlgo`` at all.
    """

    hash: str
    algo: HashAlgo | None = None
    mode: HashMode = "flat"

    def __post_init__(self) -> None:
        if self.mode not in ("flat", "recursive"):
            raise InvalidDerivationError(f"unknown hash mode {self.mode!r}")
        self.resolved()  # decode eagerly, so a bad hash fails here

    def resolved(self) -> tuple[HashAlgo, str]:
        """``(algorithm, hex digest)``, decoded from whatever was written.

        Accepts hex, Nix base-32, base-64 and SRI. The corpus contains SRI and
        base-32 and no hex at all, so an implementation that accepts only hex
        parses nothing real.
        """
        text = self.hash
        if "-" in text:
            prefix, _, body = text.partition("-")
            algo = _as_algo(prefix)
            if self.algo is not None and self.algo != algo:
                raise InvalidDerivationError(
                    f"algo={self.algo!r} contradicts the SRI prefix {prefix!r}"
                )
            raw = _b64(body)
        else:
            if self.algo is None:
                raise InvalidDerivationError(
                    "algo is required unless the hash is SRI (sha256-...)"
                )
            # Checked rather than trusted: Literal is erased at runtime, so a
            # caller who is not type-checked can still hand over "sha3".
            algo = _as_algo(self.algo)
            size = _DIGEST_BYTES[algo]
            if len(text) == size * 2:
                try:
                    raw = bytes.fromhex(text)
                except ValueError as e:
                    raise InvalidDerivationError(f"not hex: {text!r}") from e
            elif len(text) == base32_length(size):
                try:
                    raw = base32_decode(text, size)
                except ValueError as e:
                    raise InvalidDerivationError(str(e)) from e
            else:
                raw = _b64(text)
        if len(raw) != _DIGEST_BYTES[algo]:
            raise InvalidDerivationError(
                f"{algo} needs {_DIGEST_BYTES[algo]} bytes, got {len(raw)}"
            )
        return algo, raw.hex()

    @property
    def hash_algo_field(self) -> str:
        """The serialized algorithm, with the ``r:`` prefix when recursive."""
        algo, _ = self.resolved()
        return f"r:{algo}" if self.mode == "recursive" else algo

    @property
    def hash_field(self) -> str:
        """The serialized hash: always lowercase hex."""
        return self.resolved()[1]

    def env(self) -> dict[str, str]:
        """The env entries Nix synthesizes for a fixed-output derivation.

        ``outputHashAlgo`` is omitted for an SRI hash, matching what real Nix
        emits; writing it anyway would change the bytes.
        """
        out = {"outputHash": self.hash, "outputHashMode": self.mode}
        if self.algo is not None:
            out["outputHashAlgo"] = self.algo
        return out


def json_env(attrs: Mapping[str, JsonValue]) -> str:
    """Serialize attributes the way `__structuredAttrs` does.

    Canonical, and the form is measured rather than chosen: object keys sorted
    ascending, compact separators, and non-ASCII emitted raw rather than
    ``\\uXXXX`` escaped. All three hold on 456 of 456 structured derivations
    in a real closure. See ``spec/canonical.md`` section 1.8.
    """
    return json.dumps(
        attrs, sort_keys=True, separators=(",", ":"), ensure_ascii=False
    )


def _b64(text: str) -> bytes:
    try:
        return base64.b64decode(text, validate=True)
    except (binascii.Error, ValueError) as e:
        raise InvalidDerivationError(f"not base-64: {text!r}") from e


@dataclass(frozen=True, slots=True)
class Dep:
    """An edge: a derivation, and the outputs of it actually needed.

    The outputs belong to the EDGE rather than to the target, because two
    dependents of one package routinely need different outputs of it, and
    depending on ``dev`` alone is the common case in nixpkgs.
    """

    drv: Drv
    outputs: tuple[OutputName, ...] = (_DEFAULT_OUTPUT,)


@dataclass(frozen=True, slots=True)
class Drv:
    """A described derivation, and everything derivable from it.

    ``input_hash`` is the hash by which this derivation is known when it is
    someone ELSE's input, which is not the hash used to compute its own output
    paths. Keeping both on the value is what lets a dependent be built without
    re-walking the graph, and keeping them NAMED apart is what stops them being
    swapped, which is a bug this repository has already paid for.
    """

    derivation: Derivation
    path: StorePath
    input_hash: Sha256Hex

    @property
    def name(self) -> str:
        """The derivation name, as it appears in store path suffixes."""
        return self.derivation.name

    @property
    def outputs(self) -> Mapping[OutputName, StorePath]:
        """Every output name mapped to the path it will occupy."""
        return {o.name: o.path for o in self.derivation.outputs}

    def output(self, name: str = "out") -> StorePath:
        """The path of one output, known before anything is built."""
        try:
            return self.outputs[OutputName(name)]
        except KeyError:
            raise InvalidDerivationError(
                f"{self.name!r} has no output {name!r}; "
                f"it has {sorted(self.outputs)}"
            ) from None

    def ref(self, *outputs: str) -> Dep:
        """An edge to this derivation, needing ``outputs`` (default ``out``)."""
        names = tuple(OutputName(o) for o in outputs) or (_DEFAULT_OUTPUT,)
        for n in names:
            self.output(n)  # fail here rather than at build time
        return Dep(self, tuple(sorted(set(names))))

    def aterm(self) -> str:
        """The canonical bytes.

        These ARE the artifact: ``path`` is their hash, so a difference of
        one separator is a different derivation.
        """
        return unparse(self.derivation)

    def write(self, directory: pathlib.Path) -> pathlib.Path:
        """Write the ``.drv`` under ``directory``, named as in the store.

        No trailing newline: the store object does not have one, and adding one
        would change the hash of anything that reads it back.
        """
        target = directory / self.path.split("/")[-1]
        target.write_text(self.aterm())
        return target


def canonical(drv: Derivation) -> Derivation:
    """Put a derivation into canonical form.

    The orderings, all of which are load-bearing (``spec/canonical.md``):
    outputs by name, env by key, ``inputDrvs`` by store path with each inner
    name list sorted, ``inputSrcs`` ascending. ``args`` keeps its order,
    because there it is the meaning.

    This is idempotent, and it is the IDENTITY on every derivation real Nix
    emits, which is the sense in which the form is canonical rather than merely
    ours. Both are property-tested.
    """
    return Derivation(
        outputs=tuple(sorted(drv.outputs, key=lambda o: o.name)),
        input_drvs=tuple(
            sorted(
                (
                    InputDrv(i.path, tuple(sorted(set(i.outputs))))
                    for i in drv.input_drvs
                ),
                key=lambda i: i.path,
            )
        ),
        input_srcs=tuple(sorted(set(drv.input_srcs))),
        system=drv.system,
        builder=drv.builder,
        args=drv.args,
        env=tuple(sorted(drv.env)),
    )


def _edges(refs: Sequence[Dep | Drv]) -> tuple[Dep, ...]:
    """Coerce bare derivations to ``out`` edges, and merge duplicate targets.

    Two references to the same derivation are one entry with the union of the
    outputs needed, which is what Nix emits: store paths are unique within
    ``inputDrvs`` in 1293 of 1293 real derivations.
    """
    merged: dict[StorePath, tuple[Drv, set[OutputName]]] = {}
    for r in refs:
        dep = r.ref() if isinstance(r, Drv) else r
        _, names = merged.setdefault(dep.drv.path, (dep.drv, set()))
        names.update(dep.outputs)
    return tuple(
        Dep(drv, tuple(sorted(names))) for drv, names in merged.values()
    )


def derivation(
    *,
    name: str,
    system: str,
    builder: str,
    args: Sequence[str] = (),
    env: Mapping[str, JsonValue] | None = None,
    outputs: Sequence[str] | None = None,
    structured_attrs: bool = False,
    input_drvs: Sequence[Dep | Drv] = (),
    input_srcs: Sequence[str] = (),
    fixed_output: FixedOutput | None = None,
) -> Drv:
    """Describe a build. This is the whole eDSL.

    Keyword-only on purpose: nine positional arguments of mostly strings is a
    transposition waiting to happen, and ``system`` and ``builder`` are both
    strings that would silently swap.

    ``structured_attrs`` selects the SECOND env encoding (see
    ``spec/canonical.md`` section 1.8): attributes are carried as one
    ``__json`` entry with their types preserved, rather than one
    string-valued variable each. 1223 of 2516 real derivations use it, and it
    is the only way to hand a builder a list, a boolean or a nested attribute
    set without flattening it to a string. With it off, every ``env`` value
    must be a ``str``.

    ``outputs`` is an OPTION rather than a list defaulting to ``["out"]``, and
    the difference is observable in the bytes: passing ``None`` produces no
    ``outputs`` env variable (as a bare ``derivation { ... }`` does), while
    passing ``["out"]`` produces ``("outputs","out")``, as every nixpkgs
    package does. They are different derivations with different store paths.
    See ``spec/canonical.md`` section 1.7.

    Raises:
        InvalidDerivationError: if any invariant in ``spec/signature.md`` fails.
    """
    if not valid_name(name):
        raise InvalidDerivationError(
            f"{name!r} is not a valid store path name (see spec/signature.md)"
        )

    declared = outputs is not None
    names = (
        tuple(OutputName(o) for o in outputs)
        if outputs is not None
        else (_DEFAULT_OUTPUT,)
    )
    if not names:
        raise InvalidDerivationError("outputs must not be empty")
    if len(set(names)) != len(names):
        raise InvalidDerivationError(f"duplicate output names in {list(names)}")
    for n in names:
        if not valid_name(n):
            raise InvalidDerivationError(f"{n!r} is not a valid output name")
    if fixed_output is not None and len(names) != 1:
        raise InvalidDerivationError(
            "a fixed-output derivation has exactly one output, "
            f"got {list(names)}"
        )

    supplied = dict(env or {})
    reserved = _ALWAYS_RESERVED | set(names)
    clashes = sorted(reserved & supplied.keys())
    if clashes:
        raise InvalidDerivationError(
            f"env keys {clashes} are derived from the other arguments; "
            "set them through name/system/builder/outputs/fixed_output"
        )
    if not structured_attrs:
        untyped = sorted(
            k for k, v in supplied.items() if not isinstance(v, str)
        )
        if untyped:
            raise InvalidDerivationError(
                f"env values {untyped} are not strings; the flat encoding can "
                "only carry strings, so pass structured_attrs=True"
            )

    # Placeholders for the output paths. The real ones are the hash of the
    # derivation that contains them, so they cannot be known until after the
    # next step, and the masked form used to compute them blanks these anyway.
    placeholders = {str(n): "" for n in names}
    synthesized: dict[str, str]
    if structured_attrs:
        # One __json entry carrying every attribute WITH ITS TYPE, plus one
        # entry per output. The output paths stay outside the JSON, which is
        # why masking needs no special case. See spec/canonical.md 1.8.
        attrs: dict[str, JsonValue] = {
            **supplied,
            "name": name,
            "system": system,
            "builder": builder,
        }
        if declared:
            attrs["outputs"] = list(names)
        if fixed_output is not None:
            attrs.update(fixed_output.env())
        supplied = {}
        synthesized = {"__json": json_env(attrs), **placeholders}
    else:
        synthesized = {
            "name": name,
            "system": system,
            "builder": builder,
            **placeholders,
        }
        if declared:
            synthesized["outputs"] = " ".join(names)
        if fixed_output is not None:
            synthesized.update(fixed_output.env())

    algo = fixed_output.hash_algo_field if fixed_output else ""
    digest = fixed_output.hash_field if fixed_output else ""

    edges = _edges(input_drvs)
    draft = canonical(
        Derivation(
            outputs=tuple(
                Output(name=n, path=StorePath(""), hash_algo=algo, hash=digest)
                for n in names
            ),
            input_drvs=tuple(InputDrv(d.drv.path, d.outputs) for d in edges),
            input_srcs=tuple(StorePath(s) for s in input_srcs),
            system=system,
            builder=builder,
            args=tuple(args),
            env=tuple(
                {**cast("dict[str, str]", supplied), **synthesized}.items()
            ),
        )
    )

    input_hashes: dict[StorePath, str] = {
        d.drv.path: d.drv.input_hash for d in edges
    }
    paths = output_paths(draft, name, input_hashes)

    final = replace(
        draft,
        outputs=tuple(replace(o, path=paths[o.name]) for o in draft.outputs),
        env=tuple(
            (k, str(paths[OutputName(k)]) if OutputName(k) in paths else v)
            for k, v in draft.env
        ),
    )

    aterm = unparse(final)
    fixed = final.fixed_output
    return Drv(
        derivation=final,
        path=drv_path(aterm, name),
        input_hash=(
            fixed_output_input_hash(fixed)
            if fixed is not None
            else sha256_hex(unparse(final, input_hashes=input_hashes))
        ),
    )

File: impl/rust/tests/edsl.rs
//! The eDSL, checked against derivations real Nix actually emitted.
//!
//! The test that matters is not "does it build a plausible record". It is:
//! DESCRIBE the same intent that produced a golden file, and demand the same
//! bytes, including the derivation's own store path. Anything less would pass
//! for an implementation with the right shape and the wrong identity.
//!
//! These are the SAME ten intents as `impl/python/tests/test_edsl.py`. That is
//! the point: two languages, one set of bytes.

use std::collections::BTreeMap;
use std::path::{Path, PathBuf};

use img_drv::{
    Build, FixedOutput, HashAlgo, HashMode, InvalidDerivation, OutputName, base32, canonical,
    derivation, valid_name,
};

const SYSTEM: &str = "x86_64-linux";
const SH: &str = "/bin/sh";

fn golden(name: &str) -> String {
    let path = Path::new(env!("CARGO_MANIFEST_DIR"))
        .join("../../docs/spec/examples")
        .join(name);
    std::fs::read_to_string(path)
        .unwrap_or_else(|e| panic!("{name}: {e}"))
        .trim_end_matches('\n')
        .to_owned()
}

fn base(name: &str) -> Build {
    Build {
        name: name.into(),
        system: SYSTEM.into(),
        builder: SH.into(),
        ..Default::default()
    }
}

// The intents live in the library, not here: `make conformance` and the
// `examples` CLI command consume the same ten, so a corpus only the tests
// could see would not be the conformance corpus.
use img_drv::examples::{corpus as intents, dep_a, dependent, hello, multi};

#[test]
fn every_golden_file_has_an_intent() {
    // A golden nobody describes is a rule nobody is testing.
    let dir = Path::new(env!("CARGO_MANIFEST_DIR")).join("../../docs/spec/examples");
    let mut on_disk: Vec<String> = std::fs::read_dir(dir)
        .expect("golden directory")
        .filter_map(Result::ok)
        .map(|e| PathBuf::from(e.file_name()))
        .filter(|p| p.extension().is_some_and(|e| e == "drv"))
        .map(|p| p.to_string_lossy().into_owned())
        .collect();
    on_disk.sort();
    let mut described: Vec<String> = intents().iter().map(|(n, _)| (*n).to_owned()).collect();
    described.sort();
    assert_eq!(on_disk, described);
}

#[test]
fn describing_the_intent_reproduces_nix_byte_for_byte() {
    // The exit test of Phase 1, in miniature. Structural equality would not do:
    // the bytes are hashed to produce the derivation's own store path.
    for (name, drv) in intents() {
        assert_eq!(drv.aterm(), golden(name), "{name}");
    }
}

#[test]
fn the_derivation_lands_at_its_own_store_path() {
    // Reproducing the CONTENT while computing the wrong path would mean the
    // text-kind store path rule is wrong, which nothing else here would catch.
    for (name, drv) in intents() {
        let base = drv.path().as_str().rsplit('/').next().expect("basename");
        assert_eq!(base, name);
    }
}

#[test]
fn output_paths_are_known_before_anything_is_built() {
    assert_eq!(
        hello().output("out").expect("out").as_str(),
        "/nix/store/mjs27ix6ig2bkbi3s3sm470vrv4lf7ic-hello"
    );
    assert!(
        multi()
            .output("dev")
            .expect("dev")
            .as_str()
            .ends_with("-multi-dev")
    );
    assert!(
        multi()
            .output("out")
            .expect("out")
            .as_str()
            .ends_with("-multi")
    );
}

#[test]
fn a_dependent_agrees_with_what_it_depends_on() {
    let a = dep_a();
    let d = dependent();
    let edges: Vec<_> = d.derivation().input_drvs.iter().map(|i| &i.path).collect();
    assert_eq!(edges, vec![a.path()]);
    assert!(d.derivation().args[1].contains(a.output("out").expect("out").as_str()));
}

// --------------------------------------------------------------------------
// outputs is an OPTION, and both cases occur in real nixpkgs
// --------------------------------------------------------------------------

#[test]
fn declaring_outputs_is_observable_in_the_bytes() {
    // A bare `derivation { ... }` emits no `outputs` env variable; a package
    // that writes `outputs = [ "out" ]` emits `("outputs","out")`. 96 of the
    // corpus's single-output derivations do the first and 605 do the second.
    let implicit = derivation(base("x")).expect("valid");
    let explicit = derivation(Build {
        outputs: Some(vec![OutputName::new("out")]),
        ..base("x")
    })
    .expect("valid");
    assert!(!implicit.aterm().contains("(\"outputs\""));
    assert!(explicit.aterm().contains("(\"outputs\",\"out\")"));
    assert_ne!(implicit.path(), explicit.path());
    assert_ne!(
        implicit.output("out").expect("out"),
        explicit.output("out").expect("out")
    );
}

#[test]
fn the_outputs_variable_keeps_declaration_order() {
    // The outputs LIST is sorted by name; the `outputs` env VARIABLE is in
    // declaration order. They differ in 575 of the 1197 real derivations that
    // declare outputs.
    let drv = multi();
    let names: Vec<&str> = drv
        .derivation()
        .outputs
        .iter()
        .map(|o| o.name.as_str())
        .collect();
    assert_eq!(names, vec!["dev", "lib", "out"]);
    let env: BTreeMap<&str, &str> = drv
        .derivation()
        .env
        .iter()
        .map(|(k, v)| (k.as_str(), v.as_str()))
        .collect();
    assert_eq!(env["outputs"], "out dev lib");
}

// --------------------------------------------------------------------------
// fixed-output derivations
// --------------------------------------------------------------------------

#[test]
fn a_hash_is_accepted_in_every_representation_nix_writes() {
    let digest: Vec<u8> = (0u8..32).collect();
    let hex: String = digest.iter().map(|b| format!("{b:02x}")).collect();
    let forms = [
        FixedOutput::new(hex.clone(), HashAlgo::Sha256),
        FixedOutput::new(base32(&digest), HashAlgo::Sha256),
        FixedOutput::sri(sri(&digest)),
    ];
    for f in &forms {
        assert_eq!(f.hash_field().expect("decodes"), hex);
    }
}

#[test]
fn the_output_path_does_not_depend_on_how_the_hash_was_written() {
    // Why every fetchurl in nixpkgs can share one cache entry: two
    // fixed-output derivations declaring the same BYTES are interchangeable
    // however differently they are expressed. The .drv path is not, because
    // the env keeps the hash verbatim.
    let digest: Vec<u8> = (0u8..32).collect();
    let hex: String = digest.iter().map(|b| format!("{b:02x}")).collect();
    let as_hex = derivation(Build {
        fixed_output: Some(FixedOutput::new(hex, HashAlgo::Sha256)),
        ..base("fetched")
    })
    .expect("valid");
    let as_sri = derivation(Build {
        fixed_output: Some(FixedOutput::sri(sri(&digest))),
        ..base("fetched")
    })
    .expect("valid");
    assert_eq!(
        as_hex.output("out").expect("out"),
        as_sri.output("out").expect("out")
    );
    assert_ne!(as_hex.path(), as_sri.path());
}

#[test]
fn recursive_ingestion_takes_a_different_path_scheme() {
    // `r:sha256` is the `source` kind with the declared hash used directly.
    // Exactly one derivation in a 226-derivation closure exercised this.
    let digest: Vec<u8> = (0u8..32).collect();
    let hex: String = digest.iter().map(|b| format!("{b:02x}")).collect();
    let flat = derivation(Build {
        fixed_output: Some(FixedOutput::new(hex.clone(), HashAlgo::Sha256)),
        ..base("f")
    })
    .expect("valid");
    let rec = derivation(Build {
        fixed_output: Some(FixedOutput::new(hex, HashAlgo::Sha256).recursive()),
        ..base("f")
    })
    .expect("valid");
    assert!(rec.aterm().contains("\"r:sha256\""));
    assert!(rec.aterm().contains("(\"outputHashMode\",\"recursive\")"));
    assert_ne!(
        flat.output("out").expect("out"),
        rec.output("out").expect("out")
    );
}

#[test]
fn an_sri_hash_emits_no_output_hash_algo() {
    // 11 of the 93 real fixed-output derivations omit it; matching matters.
    let drv = derivation(Build {
        fixed_output: Some(FixedOutput::sri(sri(&[0u8; 32]))),
        ..base("f")
    })
    .expect("valid");
    assert!(!drv.aterm().contains("outputHashAlgo"));
}

#[test]
fn the_hash_mode_is_a_type_not_a_string() {
    // Where Python needs a runtime check because `Literal` is erased, the enum
    // makes the wrong value unrepresentable. Recorded in the typing table.
    assert_eq!(HashMode::default(), HashMode::Flat);
    assert_eq!(HashAlgo::Sha256.digest_bytes(), 32);
}

// --------------------------------------------------------------------------
// invariants from spec/signature.md, enforced at construction
// --------------------------------------------------------------------------

#[test]
fn invalid_descriptions_are_rejected_at_construction() {
    let outs = |names: &[&str]| {
        Some(
            names
                .iter()
                .map(|n| OutputName::new(*n))
                .collect::<Vec<_>>(),
        )
    };

    assert_eq!(
        derivation(Build {
            outputs: Some(vec![]),
            ..base("x")
        }),
        Err(InvalidDerivation::EmptyOutputs)
    );
    assert!(matches!(
        derivation(Build {
            outputs: outs(&["out", "out"]),
            ..base("x")
        }),
        Err(InvalidDerivation::DuplicateOutputs(_))
    ));
    assert!(matches!(
        derivation(Build {
            outputs: outs(&["out", "dev"]),
            fixed_output: Some(FixedOutput::new("0".repeat(64), HashAlgo::Sha256)),
            ..base("x")
        }),
        Err(InvalidDerivation::FixedNeedsOneOutput(_))
    ));
    for key in ["name", "out", "outputHash", "system", "builder", "outputs"] {
        assert!(
            matches!(
                derivation(Build {
                    env: BTreeMap::from([(key.to_owned(), "x".to_owned())]),
                    ..base("x")
                }),
                Err(InvalidDerivation::ReservedEnvKeys(_))
            ),
            "{key} should be reserved"
        );
    }
}

#[test]
fn invalid_names_are_rejected() {
    for bad in ["", ".", "..", ".hidden", "a b", "a/b", &"x".repeat(212)] {
        assert!(!valid_name(bad), "{bad:?}");
        assert!(matches!(
            derivation(base(bad)),
            Err(InvalidDerivation::Name(_))
        ));
    }
}

#[test]
fn names_real_packages_use_are_accepted() {
    for good in [
        "hello",
        "hello-1.0",
        "a+b",
        "x_y",
        "q?",
        "n=1",
        &"x".repeat(211),
    ] {
        assert!(valid_name(good), "{good:?}");
    }
}

#[test]
fn asking_for_an_output_that_does_not_exist_fails() {
    assert!(matches!(
        hello().output("dev"),
        Err(InvalidDerivation::NoSuchOutput { .. })
    ));
    assert!(matches!(
        hello().needs(&["dev"]),
        Err(InvalidDerivation::NoSuchOutput { .. })
    ));
}

#[test]
fn bad_hashes_are_rejected() {
    // The enum already rules out an unknown ALGORITHM, which Python has to
    // check at runtime. What is left to check is the hash itself.
    assert!(
        FixedOutput::new("0".repeat(63), HashAlgo::Sha256)
            .resolved()
            .is_err()
    );
    assert!(FixedOutput::sri("0".repeat(64)).resolved().is_err());
    assert!(
        FixedOutput::sri(format!("sha512-{}", b64(&[0u8; 32])))
            .resolved()
            .is_err()
    );
    assert!(FixedOutput::sri("sha3-AAAA").resolved().is_err());
}

#[test]
fn one_edge_per_dependency_however_often_it_is_named() {
    let m = multi();
    let drv = derivation(Build {
        input_drvs: vec![
            m.needs(&["dev"]).expect("dev"),
            m.needs(&["lib"]).expect("lib"),
            m.needs(&["dev"]).expect("dev"),
        ],
        ..base("user")
    })
    .expect("valid");
    assert_eq!(drv.derivation().input_drvs.len(), 1);
    let names: Vec<&str> = drv.derivation().input_drvs[0]
        .outputs
        .iter()
        .map(|n| n.as_str())
        .collect();
    assert_eq!(names, vec!["dev", "lib"]);
}

#[test]
fn what_the_edsl_builds_is_canonical() {
    for (name, drv) in intents() {
        assert_eq!(canonical(drv.derivation()), *drv.derivation(), "{name}");
    }
}

const B64: &[u8; 64] = b"ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/";

fn b64(data: &[u8]) -> String {
    let mut out = String::new();
    for chunk in data.chunks(3) {
        let mut acc = [0u8; 3];
        acc[..chunk.len()].copy_from_slice(chunk);
        let n = u32::from_be_bytes([0, acc[0], acc[1], acc[2]]);
        for i in 0..4 {
            if i <= chunk.len() {
                out.push(B64[((n >> (18 - 6 * i)) & 0x3f) as usize] as char);
            } else {
                out.push('=');
            }
        }
    }
    out
}

fn sri(digest: &[u8]) -> String {
    format!("sha256-{}", b64(digest))
}
